src/mapping/query/function/split.rs

```diff
@@ -1,8 +1,86 @@
 use super::prelude::*;
 use bytes::Bytes;
-use lazy_static::lazy_static;
 use std::collections::BTreeMap;
-use std::ops::Deref;
+use std::convert::TryFrom;
+
+pub struct DynamicRegex {
+    pattern: String,
+    multiline: bool,
+    insensitive: bool,
+    global: bool,
+    compiled: Option<Result<regex::Regex>>,
+}
+
+impl DynamicRegex {
+    pub fn new(pattern: String, multiline: bool, insensitive: bool, global: bool) -> Self {
+        Self {
+            pattern,
+            multiline,
+            insensitive,
+            global,
+            compiled: None,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn is_global(&self) -> bool {
+        self.global
+    }
+
+    pub fn compile(&mut self) -> Result<&regex::Regex> {
+        if self.compiled.is_none() {
+            self.compiled = Some(
+                regex::RegexBuilder::new(&self.pattern)
+                    .case_insensitive(self.insensitive)
+                    .multi_line(self.multiline)
+                    .build()
+                    .map_err(|err| format!("invalid regex {}", err)),
+            );
+        }
+
+        self.compiled
+            .as_ref()
+            // We know this unwrap is safe because we have just populated the Option above.
+            .unwrap()
+            .as_ref()
+            .map_err(|err| err.to_string())
+    }
+}
+
+impl TryFrom<BTreeMap<String, Value>> for DynamicRegex {
```

Comment:
If we make this `TryFrom<Value> for DynamicRegex`, we can have both the `Value::Map` and `Value::Bytes` case be handled by this conversion, and have `DynamicRegex` be useable in both situations.

It seems like this can reduce a bunch of duplication, possibl?

Reply:
I'm wondering if we do want to create a `DynamicRegex` from a `Value::Bytes`? In a function such as `split`, if you split by a string, then that means you want to split by that exact string and not a regex.

So, 

`split("apple.pear", /./)`

will be different to

`split("apple.pear", ".")`

Reply:
No sorry, I didn’t mean creating a dynamic regex from a string. I meant using a singular type to represent both a dynamic and static regex, to allow consolidating some logic.

Basically a `CoW` type, where a static regex is used if one is stored, or else the dynamic regex is compiled to a static one (and served from then on if the regex is requested from the type).